rendezvous/README.md

```diff
@@ -186,7 +186,55 @@ register under a namespace of interest (eg: the relay namespace). This
 can be mitigated by requiring a Proof of Work scheme for client
 registrations.
 
-This is TBD before finalizing the spec.
+We define "difficulty" as the number of zero bytes at the front of the
+hash assuming a big endian encoding.
+
+If a rendezvous node decides that Proof of Work is required to complete
+a registration, it will send a response with the status code `E_POW_REQUIRED`
+together with a `challenge` and the `target_difficulty`.
+
+A client is then expected to follow up with a `ProofOfWork` message, computing
+a hash by concatenating the following:
+
+- The UTF8 bytes of the string `libp2p-rendezvous-pow`
+- The contents of `challenge`
+- The `ns` field
+- The `signedPeerRecord` field
+- The `PeerId` of the rendezvous point we are registering with
+- The `nonce` field
+
+The resulting byte buffer is hashed using SHA256.
+
+If a rendezvous point considers the difficulty too low, it will decline a
+`ProofOfWork` message with a `RegisterResponse` and the status code set to
+`E_POW_DIFFICULTY_TOO_LOW` together with a new `challenge` and the `target_difficulty`.
+
+Once the difficulty requirement is met, a rendezvous point will respond with a
+status code of `OK`.
+
+To compute the PoW, a client picks a `nonce`, hashes the above buffer and checks
+the resulting hash for its difficulty. If the requirement is met, the hash can
+be sent to the server. If the requirement is not met, a new `nonce` needs to be
+picked and the whole process is repeated. How the `nonce` is picked - be it by
+iterating from i64::MIN to i64::MAX or choosing randomly doesn't matter.
+
+Example:
+
+```
+A -> R: REGISTER{my-app, {QmA, AddrA}}
+R -> A: {E_POW_REQUIRED, challenge: 0xDEADBEEFDEADBEEF, target_difficulty: 3}
+A -> R: PROOF_OF_WORK{0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF, nonce: 3123452} (client ignores `target difficulty`)
+R -> A: {E_DIFFICULTY_TOO_LOW, challenge: 0xBEEFDEADBEEFDEAD, target_difficulty: 3}
+A -> R: PROOF_OF_WORK{0x000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF, nonce: 8572945}
+R -> A: {OK}
+```
+
+It is worth noting that the `target_difficulty` can change between the initial `E_POW_REQUIRED`
+and submitting the PoW. Henceforth, a client must be prepared to receive a `E_DIFFICULTY_TOO_LOW`
+message even if they adhered to the previously sent `target_difficulty`.
```

Comment:
Thinking that this is actually quite annoying and there isn't really a reason for the server to do this. Once they've requested a specific PoW, they SHOULD accept it if the client provides a legitimate PoW.